File: VB/CheckBoxLegendExample/CheckBoxLegend/ChartLegendHelper.vb
Imports DevExpress.DashboardWin
Imports DevExpress.XtraCharts
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text

Namespace CheckBoxLegend
    Friend Class ChartLegendHelper
        Private viewer As DashboardViewer
        Private designer As DashboardDesigner
        Private chartItemName As String

        Private chartStorage As List(Of ChartControl)
        Private uncheckedSeries As New List(Of String)()


        Public Sub Attach(ByVal viewer As DashboardViewer, ByVal chartItemName As String)
            If designer IsNot Nothing OrElse viewer IsNot Nothing Then
                Detach()
            End If
            chartStorage = New List(Of ChartControl)()
            Me.viewer = viewer
            Me.chartItemName = chartItemName
            AddHandler viewer.DashboardItemControlCreated, AddressOf DashboardItemControlCreated
            AddHandler viewer.DashboardItemControlUpdated, AddressOf DashboardItemControlUpdated
            AddHandler viewer.DashboardItemBeforeControlDisposed, AddressOf DashboardItemBeforeControlDisposed

        End Sub

        Public Sub Attach(ByVal designer As DashboardDesigner, ByVal chartItemName As String)
            If designer IsNot Nothing OrElse viewer IsNot Nothing Then
                Detach()
            End If
            chartStorage = New List(Of ChartControl)()
            Me.designer = designer
            Me.chartItemName = chartItemName
            AddHandler designer.DashboardItemControlCreated, AddressOf DashboardItemControlCreated
            AddHandler designer.DashboardItemControlUpdated, AddressOf DashboardItemControlUpdated
            AddHandler designer.DashboardItemBeforeControlDisposed, AddressOf DashboardItemBeforeControlDisposed
        End Sub

        Private Sub DashboardItemBeforeControlDisposed(ByVal sender As Object, ByVal e As DashboardItemControlEventArgs)
            If e.DashboardItemName<>chartItemName OrElse e.ChartControl Is Nothing Then
                Return
            End If

            RemoveHandler e.ChartControl.LegendItemChecked, AddressOf ChartControl_LegendItemChecked
            chartStorage.Remove(e.ChartControl)
        End Sub

        Private Sub DashboardItemControlUpdated(ByVal sender As Object, ByVal e As DashboardItemControlEventArgs)
            If e.DashboardItemName <> chartItemName OrElse e.ChartControl Is Nothing Then
                Return
            End If

            ApplyCheckedState(e.ChartControl)
        End Sub

        Private Sub ApplyCheckedState(ByVal chart As ChartControl)
            For Each series As Series In chart.Series
                series.Visible = Not uncheckedSeries.Contains(series.Name)
                series.CheckedInLegend = Not uncheckedSeries.Contains(series.LegendTextPattern)
            Next series

        End Sub

        Private Sub DashboardItemControlCreated(ByVal sender As Object, ByVal e As DashboardItemControlEventArgs)
            If e.DashboardItemName <> chartItemName OrElse e.ChartControl Is Nothing Then
                Return
            End If
            chartStorage.Add(e.ChartControl)
            e.ChartControl.Legend.MarkerMode = DevExpress.XtraCharts.LegendMarkerMode.CheckBoxAndMarker
            AddHandler e.ChartControl.LegendItemChecked, AddressOf ChartControl_LegendItemChecked
        End Sub

        Private Sub ChartControl_LegendItemChecked(ByVal sender As Object, ByVal e As DevExpress.XtraCharts.LegendItemCheckedEventArgs)
            Dim chart As ChartControl = DirectCast(sender, ChartControl)
            If TypeOf e.CheckedElement Is Series Then
                Dim series = CType(e.CheckedElement, Series)
                chart.Series(series.LegendTextPattern).Visible = e.NewCheckState
                If Not e.NewCheckState Then
                    If Not uncheckedSeries.Contains(series.LegendTextPattern) Then
                        uncheckedSeries.Add(series.LegendTextPattern)
                    End If
                ElseIf uncheckedSeries.Contains(series.LegendTextPattern) Then
                    uncheckedSeries.Remove(series.LegendTextPattern)
                End If
            End If

            ' Update minimized chart
            If chartStorage.Count > 1 Then
            For Each chartControl As ChartControl In chartStorage
                If chart Is chartControl Then
                    Continue For
                End If
                    ApplyCheckedState(chartControl)
            Next chartControl
            End If
        End Sub


        Public Sub Detach()
            Me.chartItemName = String.Empty
            If Me.designer IsNot Nothing Then
                RemoveHandler designer.DashboardItemControlCreated, AddressOf DashboardItemControlCreated
                RemoveHandler designer.DashboardItemControlUpdated, AddressOf DashboardItemControlUpdated
                RemoveHandler designer.DashboardItemBeforeControlDisposed, AddressOf DashboardItemBeforeControlDisposed
                Me.designer = Nothing
            End If
            If Me.viewer IsNot Nothing Then
                RemoveHandler viewer.DashboardItemControlCreated, AddressOf DashboardItemControlCreated
                RemoveHandler viewer.DashboardItemControlUpdated, AddressOf DashboardItemControlUpdated
                RemoveHandler viewer.DashboardItemBeforeControlDisposed, AddressOf DashboardItemBeforeControlDisposed
                Me.viewer = Nothing
            End If
                chartStorage = Nothing


        End Sub
    End Class
End Namespace

File: CS/CheckBoxLegendExample/CheckBoxLegend/ChartLegendHelper.cs
using DevExpress.DashboardWin;
using DevExpress.XtraCharts;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace CheckBoxLegend
{
    class ChartLegendHelper
    {
        DashboardViewer viewer;
        DashboardDesigner designer;
        string chartItemName;

        List<ChartControl> chartStorage;
        List<string> uncheckedSeries = new List<string>();


        public void Attach( DashboardViewer viewer, string chartItemName)
        {
            if (designer != null || viewer != null)
                Detach();
            chartStorage = new List<ChartControl>();
            this.viewer = viewer;
            this.chartItemName = chartItemName;
            viewer.DashboardItemControlCreated += DashboardItemControlCreated;
            viewer.DashboardItemControlUpdated += DashboardItemControlUpdated;
            viewer.DashboardItemBeforeControlDisposed += DashboardItemBeforeControlDisposed;

        }

        public void Attach(DashboardDesigner designer, string chartItemName)
        {
            if (designer != null || viewer != null)
                Detach();
            chartStorage = new List<ChartControl>();
            this.designer = designer;
            this.chartItemName = chartItemName;
            designer.DashboardItemControlCreated += DashboardItemControlCreated;
            designer.DashboardItemControlUpdated += DashboardItemControlUpdated;
            designer.DashboardItemBeforeControlDisposed += DashboardItemBeforeControlDisposed;
        }

        private void DashboardItemBeforeControlDisposed(object sender, DashboardItemControlEventArgs e)
        {
            if (e.DashboardItemName!=chartItemName || e.ChartControl == null) return;

            e.ChartControl.LegendItemChecked -= ChartControl_LegendItemChecked;
            chartStorage.Remove(e.ChartControl);
        }

        private void DashboardItemControlUpdated(object sender, DashboardItemControlEventArgs e)
        {
            if (e.DashboardItemName != chartItemName || e.ChartControl == null) return;

            ApplyCheckedState(e.ChartControl);
        }

        private void ApplyCheckedState(ChartControl chart)
        {
            foreach (Series series in chart.Series)
            {
                series.Visible = !uncheckedSeries.Contains(series.Name);
                series.CheckedInLegend = !uncheckedSeries.Contains(series.LegendTextPattern);
            }

        }

        private void DashboardItemControlCreated(object sender, DashboardItemControlEventArgs e)
        {
            if (e.DashboardItemName != chartItemName || e.ChartControl == null) return;
            chartStorage.Add(e.ChartControl);
            e.ChartControl.Legend.MarkerMode = DevExpress.XtraCharts.LegendMarkerMode.CheckBoxAndMarker;
            e.ChartControl.LegendItemChecked += ChartControl_LegendItemChecked;
        }

        private void ChartControl_LegendItemChecked(object sender, DevExpress.XtraCharts.LegendItemCheckedEventArgs e)
        {
            ChartControl chart = (ChartControl)sender;
            if (e.CheckedElement is Series)
            {
                var series = (Series)e.CheckedElement;
                chart.Series[series.LegendTextPattern].Visible = e.NewCheckState;
                if (!e.NewCheckState)
                {
                    if (!uncheckedSeries.Contains(series.LegendTextPattern))
                        uncheckedSeries.Add(series.LegendTextPattern);
                }
                else if (uncheckedSeries.Contains(series.LegendTextPattern))
                    uncheckedSeries.Remove(series.LegendTextPattern);
            }

            // Update minimized chart
            if ( chartStorage.Count > 1)
            foreach(ChartControl chartControl in chartStorage)
            {
                if (chart == chartControl) continue;
                    ApplyCheckedState(chartControl);
            }
        }


        public void Detach()
        {
            this.chartItemName = string.Empty;
            if (this.designer != null)
            {
                designer.DashboardItemControlCreated -= DashboardItemControlCreated;
                designer.DashboardItemControlUpdated -= DashboardItemControlUpdated;
                designer.DashboardItemBeforeControlDisposed -= DashboardItemBeforeControlDisposed;
                this.designer = null;
            }
            if (this.viewer != null)
            {
                viewer.DashboardItemControlCreated -= DashboardItemControlCreated;
                viewer.DashboardItemControlUpdated -= DashboardItemControlUpdated;
                viewer.DashboardItemBeforeControlDisposed -= DashboardItemBeforeControlDisposed;
                this.viewer = null;
            }
                chartStorage = null;


        }
    }
}
